File: common/src/main/java/generations/gg/generations/core/generationscore/client/model/RunnableKeybind.java
package generations.gg.generations.core.generationscore.client.model;

import dev.architectury.event.EventResult;
import dev.architectury.event.events.client.ClientRawInputEvent;
import dev.architectury.registry.client.keymappings.KeyMappingRegistry;
import net.minecraft.client.KeyMapping;
import net.minecraft.client.Minecraft;
import org.lwjgl.glfw.GLFW;

public class RunnableKeybind extends KeyMapping {

    private final Runnable runnable;

    public static void create(String name, int glfwKey, String category, Runnable action) {
        RunnableKeybind keybind = new RunnableKeybind(name, glfwKey, category, action);
        KeyMappingRegistry.register(keybind);
    }

    public RunnableKeybind(String name, int keyId, String category, Runnable runnable) {
        super(name, keyId, category);
        this.runnable = runnable;
        ClientRawInputEvent.KEY_PRESSED.register((client, keyCode, scanCode, action1, modifiers) -> onKeyPress(client, keyCode, scanCode, action1, modifiers));
    }

    public EventResult onKeyPress(Minecraft client, int keyCode, int scanCode, int action, int modifiers) {
        if (action == GLFW.GLFW_RELEASE && keyCode == this.getDefaultKey().getValue()) {
            if(this.consumeClick()) runnable.run();
            return EventResult.interruptFalse();
        }

        return EventResult.pass();
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/client/model/RunnableKeybind.kt
package generations.gg.generations.core.generationscore.common.client.model

import generations.gg.generations.core.generationscore.common.client.render.rarecandy.Pipelines
import net.minecraft.client.KeyMapping
import org.lwjgl.glfw.GLFW

object Keybinds {
    val keybindings = mutableListOf<RunnableKeybind>()

    val TOGGLE_RENDERING = register("toggleShaderRendering", GLFW.GLFW_KEY_P, "rendering", Pipelines::toggleRendering)

    private fun register(name: String, keyId: Int, category: String, runnable: Runnable) = RunnableKeybind(name, keyId, category, runnable).also(keybindings::add)

    fun pressDown(keyCode: Int, scanCode: Int, action: Int, modifiers: Int): Boolean {
        for(keybinding in keybindings) {
            if(keybinding.onKeyPress(keyCode, scanCode, action, modifiers)) return true
        }

        return false
    }

    class RunnableKeybind(name: String, keyId: Int, category: String, private val runnable: Runnable) : KeyMapping(name, keyId, category) {
        fun onKeyPress(keyCode: Int, scanCode: Int, action: Int, modifiers: Int): Boolean {
            if (action == GLFW.GLFW_RELEASE && keyCode == defaultKey.value) {
                if (this.consumeClick()) runnable.run()
                return true
            }

            return false
        }
    }
}
